File: pvm/ts/src/engine/managers/dealerManager.ts
import { PlayerStatus } from "@block52/poker-vm-sdk";
import { Player } from "../../models/player";
import { IDealerGameInterface, IDealerPositionManager } from "../types";

/**
 * Standalone dealer position manager that works with TexasHoldemGame instances
 */
export class DealerPositionManager implements IDealerPositionManager {
    private readonly game: IDealerGameInterface;

    constructor(game: IDealerGameInterface) {
        this.game = game;
    }

    /**
     * Rotates dealer position to the next active player after each hand
     */
    private rotateDealer(): number {
        // Use effective dealer position (the actual player who was dealer)
        const currentDealer = this.getEffectiveDealerPosition();
        const activePlayers = this.game.findActivePlayers();

        // Find the next active player after the current dealer
        const nextDealer = this.findNextActivePlayer(currentDealer);

        if (nextDealer) {
            const nextDealerSeat = this.game.getPlayerSeatNumber(nextDealer.address);
            console.log(`Rotating dealer from seat ${currentDealer} to seat ${nextDealerSeat}`);
            return nextDealerSeat;
        }

        // Fallback: if no next player found, start from first active player
        const firstActivePlayer = activePlayers[0];
        const fallbackSeat = this.game.getPlayerSeatNumber(firstActivePlayer.address);
        return fallbackSeat;
    }

    public getPosition(name: string): number {
        switch (name.toLowerCase()) {
            case "dealer":
            case "d":
                return this.getDealerPosition();
            case "small blind":
            case "small-blind":
            case "sb":
                return this.getSmallBlindPosition();
            case "big blind":
            case "big-blind":
            case "bb":
                return this.getBigBlindPosition();
            default:
                throw new Error(`Unknown position name: ${name}`);
        }
    }
    /**
     * Gets the current dealer position, initializing if necessary
     */
    public getDealerPosition(): number {
        const currentDealerSeat = this.getCurrentDealerSeat();
        return currentDealerSeat;
    }

    /**
     * Finds the next active player starting from a given seat
     */
    public findNextActivePlayer(startSeat: number): Player | undefined {
        const maxSeats = this.game.maxPlayers;

        // Start searching from the next seat after startSeat
        let searchSeat = startSeat + 1;
        if (searchSeat > maxSeats) {
            searchSeat = 1;
        }

        // Search from startSeat+1 to maxSeats
        for (let seat = searchSeat; seat <= maxSeats; seat++) {
            const player = this.game.getPlayerAtSeat(seat);
            if (player && this.isPlayerActive(player)) {
                return player;
            }
        }

        // Wrap around and search from 1 to startSeat
        for (let seat = 1; seat < startSeat; seat++) {
            const player = this.game.getPlayerAtSeat(seat);
            if (player && this.isPlayerActive(player)) {
                return player;
            }
        }

        return undefined;
    }

    /**
     * Finds the next active player starting from a given seat
     */
    private findNextPlayer(startSeat: number): Player | undefined {
        const maxSeats = this.game.maxPlayers;

        // Start searching from the next seat after startSeat
        let searchSeat = startSeat + 1;
        if (searchSeat > maxSeats) {
            searchSeat = 1;
        }

        // Search from startSeat+1 to maxSeats
        for (let seat = searchSeat; seat <= maxSeats; seat++) {
            const player = this.game.getPlayerAtSeat(seat);
            if (player && this.isPlayerSittingIn(player)) {
                return player;
            }
        }

        // Wrap around and search from 1 to startSeat
        for (let seat = 1; seat < startSeat; seat++) {
            const player = this.game.getPlayerAtSeat(seat);
            if (player && this.isPlayerSittingIn(player)) {
                return player;
            }
        }

        return undefined;
    }

    /**
     * Handles dealer position when a player leaves
     */
    public handlePlayerLeave(seat: number): void {
        const currentDealerSeat = this.getCurrentDealerSeat();

        // If the leaving player is the dealer, rotate to next player
        if (currentDealerSeat === seat) {
            const remainingPlayers = this.game.findActivePlayers().filter(p => this.game.getPlayerSeatNumber(p.address) !== seat);

            if (remainingPlayers.length >= this.game.minPlayers) {
                this.rotateDealer();
            } else if (remainingPlayers.length === 1) {
                // Only one player left, they become dealer
                // const newDealerSeat = this.game.getPlayerSeatNumber(remainingPlayers[0].address);
                // this.setDealerPosition(newDealerSeat);
            }
        }
    }

    /**
     * Handles dealer position when a new player joins
     */
    public handlePlayerJoin(_seat: number): void {
        // const activePlayers = this.game.findActivePlayers();

        // if (activePlayers.length === 1) {
        //     // First player becomes dealer by default
        //     // this.setDealerPosition(newPlayerSeat);
        // } else if (activePlayers.length === this.game.minPlayers && !this.getCurrentDealerSeat()) {
        //     // If somehow dealer wasn't set, initialize it
        // }
        // // For more players, dealer position doesn't change when someone joins
    }

    /**
     * Validates dealer position consistency
     */
    public validateDealerPosition(): boolean {
        const dealerSeat = this.getCurrentDealerSeat();

        if (!dealerSeat) {
            return false;
        }

        const dealerPlayer = this.game.getPlayerAtSeat(dealerSeat);
        if (!dealerPlayer) {
            return false;
        }

        return this.isPlayerActive(dealerPlayer);
    }

    /**
     * Gets small blind position based on dealer
     */
    public getSmallBlindPosition(): number {
        const dealerSeat = this.getDealerPosition();
        const activePlayers = this.game.findActivePlayers();

        // Heads-up: effective dealer is the small blind
        if (activePlayers.length === 2) {
            return this.getEffectiveDealerPosition();
        }

        // 3+ players: small blind is next active player after dealer
        const sbPlayer = this.findNextActivePlayer(dealerSeat);
        return sbPlayer ? this.game.getPlayerSeatNumber(sbPlayer.address) : (dealerSeat % this.game.maxPlayers) + 1;
    }

    /**
     * Gets big blind position based on dealer and small blind
     */
    public getBigBlindPosition(): number {
        const activePlayers = this.game.findActivePlayers();

        // Heads-up: big blind is the non-dealer (next player after effective dealer)
        if (activePlayers.length === 2) {
            const effectiveDealer = this.getEffectiveDealerPosition();
            const bbPlayer = this.findNextActivePlayer(effectiveDealer);
            return bbPlayer ? this.game.getPlayerSeatNumber(bbPlayer.address) : (effectiveDealer % this.game.maxPlayers) + 1;
        }

        // 3+ players: big blind is next active player after small blind
        const sbSeat = this.getSmallBlindPosition();
        const bbPlayer = this.findNextActivePlayer(sbSeat);
        return bbPlayer ? this.game.getPlayerSeatNumber(bbPlayer.address) : (sbSeat % this.game.maxPlayers) + 1;
    }

    /**
     * Helper method to check if a player is active
     */
    private isPlayerActive(player: Player): boolean {
        return (player.status === PlayerStatus.ACTIVE || player.status === PlayerStatus.SHOWING) && player.chips > 0;
    }

    private isPlayerSittingIn(player: Player): boolean {
        return player.status !== PlayerStatus.SITTING_OUT;
    }

    /**
     * Helper method to get current dealer seat from game
     */
    getCurrentDealerSeat(): number {
        // Access the dealer position from the game's positions object
        return this.game.dealerPosition;
    }

    /**
     * Gets the effective dealer position (the actual player holding the button)
     * If dealer seat is empty, finds the next active player after dealer button
     */
    private getEffectiveDealerPosition(): number {
        const dealerSeat = this.getDealerPosition();
        const dealerPlayer = this.game.getPlayerAtSeat(dealerSeat);

        // If there's an active player at the dealer seat, they're the dealer
        if (dealerPlayer && this.isPlayerActive(dealerPlayer)) {
            return dealerSeat;
        }

        // Otherwise, find the next active player after the dealer button
        const nextPlayer = this.findNextActivePlayer(dealerSeat);
        return nextPlayer ? this.game.getPlayerSeatNumber(nextPlayer.address) : dealerSeat;
    }

    /**
     * Method to handle new hand initialization
     */
    public handleNewHand(): number {
        return this.rotateDealer();
        // const activePlayers = this.game.findActivePlayers();

        // if (activePlayers.length === 2) {
        //     // Heads-up: dealer alternates each hand
        //     return this.handleHeadsUpDealer();
        // } else {
        //     // Multi-player: rotate to next active player
        //     return this.rotateDealer();
        // }
    }
}

File: pvm/haskell/src/TexasHoldem/DealerManager.hs
{-|
Module      : TexasHoldem.DealerManager
Description : Dealer button and position management
Copyright   : (c) Block52, 2025
License     : MIT
Maintainer  : dev@block52.xyz

This module manages dealer button position and calculates blind positions.
Handles heads-up special case where dealer posts small blind.
Compatible with the TypeScript DealerPositionManager implementation.
-}
module TexasHoldem.DealerManager
    ( -- * Types
      DealerState(..)
    , Position(..)
      -- * Constructors
    , mkDealerState
      -- * Position Queries
    , getDealerPosition
    , getSmallBlindPosition
    , getBigBlindPosition
    , getPosition
    , getEffectiveDealerPosition
      -- * Player Finding
    , findNextActivePlayer
    , findNextPlayer
      -- * State Updates
    , rotateDealer
    , handleNewHand
    , handlePlayerJoin
    , handlePlayerLeave
      -- * Validation
    , validateDealerPosition
      -- * Utilities
    , isHeadsUp
    ) where

import TexasHoldem.Player

import qualified Data.Map.Strict as Map
import Data.List (find)
import Data.Maybe (isJust, fromMaybe)

-- | Named positions at the table
data Position = Dealer | SmallBlind | BigBlind
    deriving (Eq, Show, Read, Ord, Enum, Bounded)

-- | Dealer management state
data DealerState = DealerState
    { dsDealerSeat   :: !SeatIndex           -- ^ Current dealer button position
    , dsMaxSeats     :: !Int                 -- ^ Maximum seats at table
    , dsMinPlayers   :: !Int                 -- ^ Minimum players to start
    } deriving (Eq, Show)

-- | Create initial dealer state
mkDealerState :: SeatIndex -> Int -> Int -> DealerState
mkDealerState dealerSeat maxSeats minPlayers = DealerState
    { dsDealerSeat = dealerSeat
    , dsMaxSeats   = maxSeats
    , dsMinPlayers = minPlayers
    }

-- | Get current dealer position
getDealerPosition :: DealerState -> SeatIndex
getDealerPosition = dsDealerSeat

-- | Get position by name
getPosition :: Position -> DealerState -> Map.Map SeatIndex Player -> SeatIndex
getPosition pos ds players = case pos of
    Dealer     -> getDealerPosition ds
    SmallBlind -> getSmallBlindPosition ds players
    BigBlind   -> getBigBlindPosition ds players

-- | Get small blind position based on dealer and number of players
-- Heads-up: dealer is small blind
-- 3+ players: next active player after dealer
getSmallBlindPosition :: DealerState -> Map.Map SeatIndex Player -> SeatIndex
getSmallBlindPosition ds players
    | isHeadsUp players = getEffectiveDealerPosition ds players
    | otherwise = case findNextActivePlayer (dsDealerSeat ds) ds players of
        Just p  -> playerSeat p
        Nothing -> nextSeat (dsDealerSeat ds) (dsMaxSeats ds)

-- | Get big blind position based on small blind
-- Heads-up: next player after effective dealer
-- 3+ players: next active player after small blind
getBigBlindPosition :: DealerState -> Map.Map SeatIndex Player -> SeatIndex
getBigBlindPosition ds players
    | isHeadsUp players =
        let effectiveDealer = getEffectiveDealerPosition ds players
        in case findNextActivePlayer effectiveDealer ds players of
            Just p  -> playerSeat p
            Nothing -> nextSeat effectiveDealer (dsMaxSeats ds)
    | otherwise =
        let sbSeat = getSmallBlindPosition ds players
        in case findNextActivePlayer sbSeat ds players of
            Just p  -> playerSeat p
            Nothing -> nextSeat sbSeat (dsMaxSeats ds)

-- | Get effective dealer position (actual player holding button)
-- If dealer seat is empty, finds next active player
getEffectiveDealerPosition :: DealerState -> Map.Map SeatIndex Player -> SeatIndex
getEffectiveDealerPosition ds players =
    case Map.lookup (dsDealerSeat ds) players of
        Just p | isActive p -> dsDealerSeat ds
        _ -> case findNextActivePlayer (dsDealerSeat ds) ds players of
            Just p  -> playerSeat p
            Nothing -> dsDealerSeat ds

-- | Find next active player clockwise from given seat
findNextActivePlayer :: SeatIndex -> DealerState -> Map.Map SeatIndex Player -> Maybe Player
findNextActivePlayer startSeat ds players = findNextPlayerWith isActive startSeat ds players

-- | Find next player (not sitting out) clockwise from given seat
findNextPlayer :: SeatIndex -> DealerState -> Map.Map SeatIndex Player -> Maybe Player
findNextPlayer startSeat ds players = findNextPlayerWith (not . isSittingOut) startSeat ds players
  where
    isSittingOut p = playerStatus p == Waiting

-- | Generic function to find next player matching predicate
findNextPlayerWith :: (Player -> Bool) -> SeatIndex -> DealerState -> Map.Map SeatIndex Player -> Maybe Player
findNextPlayerWith predicate startSeat ds players =
    let maxSeats = dsMaxSeats ds
        -- Search seats from startSeat+1 to maxSeats, then 0 to startSeat
        searchOrder = [nextSeat startSeat maxSeats .. maxSeats - 1]
                   ++ [0 .. startSeat - 1]
        findAtSeat seat = Map.lookup seat players >>= \p ->
            if predicate p then Just p else Nothing
    in foldr (\seat acc -> case acc of
                Nothing -> findAtSeat seat
                just    -> just) Nothing searchOrder

-- | Rotate dealer to next active player
rotateDealer :: DealerState -> Map.Map SeatIndex Player -> DealerState
rotateDealer ds players =
    case findNextActivePlayer (getEffectiveDealerPosition ds players) ds players of
        Just p  -> ds { dsDealerSeat = playerSeat p }
        Nothing -> ds  -- No change if no active players found

-- | Handle new hand - rotates dealer
handleNewHand :: DealerState -> Map.Map SeatIndex Player -> DealerState
handleNewHand = rotateDealer

-- | Handle player joining (dealer doesn't change)
handlePlayerJoin :: SeatIndex -> DealerState -> Map.Map SeatIndex Player -> DealerState
handlePlayerJoin _ ds _ = ds  -- Dealer position doesn't change on join

-- | Handle player leaving
-- If leaving player is dealer, rotate to next player
handlePlayerLeave :: SeatIndex -> DealerState -> Map.Map SeatIndex Player -> DealerState
handlePlayerLeave leavingSeat ds players
    | dsDealerSeat ds == leavingSeat = rotateDealer ds remainingPlayers
    | otherwise = ds
  where
    remainingPlayers = Map.delete leavingSeat players

-- | Validate that dealer position is valid
validateDealerPosition :: DealerState -> Map.Map SeatIndex Player -> Bool
validateDealerPosition ds players =
    case Map.lookup (dsDealerSeat ds) players of
        Nothing -> False
        Just p  -> isActive p

-- | Check if we're in heads-up (2 active players)
isHeadsUp :: Map.Map SeatIndex Player -> Bool
isHeadsUp players = countActivePlayers players == 2

-- | Count active players
countActivePlayers :: Map.Map SeatIndex Player -> Int
countActivePlayers = length . filter isActive . Map.elems

-- | Get next seat clockwise (wrapping around)
nextSeat :: SeatIndex -> Int -> SeatIndex
nextSeat current maxSeats = (current + 1) `mod` maxSeats
